Shared memory changes for castanets -3

Suhaspoornachandra opened the pull request:
Reference:

1. https://github.com/Samsung/Castanets/pull/282/commits/08fdaa7b12d347c4e6cab8e3ee1d704b6d887896
2. https://github.com/Samsung/Castanets/pull/282/commits/221cfe266df2e4de2ab553e76a881c8c10b6f133
3. https://github.com/Samsung/Castanets/pull/282/commits/de26b9ba7a907b32908e4c218f25332f0039b3f7
4. https://github.com/Samsung/Castanets/pull/282/commits/f5ad5ee41b751ac14601f1a54069acf6c91923ac
5. https://github.com/Samsung/Castanets/pull/282/commits/2eef176c2b8b604321a2c12566beedd48942a0ff
6. https://github.com/Samsung/Castanets/pull/282/commits/1e70559c482dff4a34a97314f0285e6a8e90c1b7
7. https://github.com/Samsung/Castanets/pull/282/commits/75c8526209d65f6d92cc01380ab88fcfa7b58651

Changes from https://github.com/Samsung/Castanets/pull/282/commits/9905ae6c109c8ec785887ad821106313defbd216 are applied in the commits of "Add SharedMemoryLocker & AutoGuidLock" and "Apply the fence to the sync message", as they were fixup of index increment.

Commit:
Add CastanetsMemoryHolder class.

- Call AddHolder() in CreateAnonymousSharedMemoryIfNeeded().
- Call RemoveHolder() in SharedMemoryTracker::IncrementMemoryUsage().
- Hold the holder until at least one mapping information is added.
- Hold the mapped memory with the same life time as CastanetsMemoryHolder.

Signed-off-by: Suhaspoornachandra <s.poornachan@samsung.com>

Commit:
Add SharedMemoryLocker & AutoGuidLock.

This CL introduces locks by GUID which
1) prevent CastanetsMemoryHolder from being created or deleted
2) prevent the addition or removal of CastanetsMemoryMapping
while the shared memory performs data copy in OnBufferSync().

Signed-off-by: Suhaspoornachandra <s.poornachan@samsung.com>

Commit:
Add CastanetsFence|Queue|Manager classes.

Castanets will introduce a new fence method to access shared memory
and sync data. This CL adds the base classes for this.

- CastanetsFence: Each BUFFER_SYNC message is paired with one Fence,
  which has a WaitableEvent as a member variable.
- CastanetsFenceQueue: Each node has one FenceQueue.
  When you receive a BUFFER_SYNC message, create a Fence and push it.
  When you're done syncing, pop the Fence to remove it.
- CastanetsFenceManager: The above FenceQueues notify the manager
  whenever a Fence is pushed or poped. The manager manages
  FenceQueues based on GUID.

Signed-off-by: Suhaspoornachandra <s.poornachan@samsung.com>

Commit:
Apply the fence to the sync message.

- Add ADD_SYNC_FENCE message paired with BUFFER_SYNC message.
- Add FenceId attribute to BUFFER_SYNC message
- BUFFER_SYNC and ADD_SYNC_FENCE must be sent together
  with the same fence id.
- BUFFER_SYNC is sent to the broker channel.
  ADD_SYNC_FENCE is sent to the node channel.

- NodeChannel has ownership of FenceManager
- BrokerCastanets has a FenceQueue. It also has a reference
  to the node channel in order to send ADD_SYNC_FENCE.

- When an unknown shared memory is converted to external memory,
  it has to send the pending syncs immediately. To do this,
  add a write method with no lock to channel_posix.

Signed-off-by: Suhaspoornachandra <s.poornachan@samsung.com>

Commit:
Add a WaitSync function and apply it to the loader and bitmap.

This CL adds a mojo function that waits for the sync to complete.
Wait/Signal are done using the fence method introduced in previous
commits. Currently applied to the loader and bitmap, we should also
add WaitSync calls properly to modules that use shared memory

Signed-off-by: Suhaspoornachandra <s.poornachan@samsung.com>

Commit:
[DCHECK] Allow WaitSync method to wait on all threads.

FATAL:thread_restrictions.cc(105)] Check failed:
!g_base_sync_primitives_disallowed.Get().Get().
Waiting on a //base sync primitive is not allowed on this thread
to prevent jank and deadlock. If waiting on a //base sync primitive
is unavoidable, do it within the scope of a ScopedAllowBaseSyncPrimitives.
If in a test use ScopedAllowBaseSyncPrimitivesForTesting.

0 0x7ff90950b10c base::debug::StackTrace::StackTrace()
1 0x7ff90943323b logging::LogMessage::~LogMessage()
2 0x7ff9094d6f76 base::internal::AssertBaseSyncPrimitivesAllowed()
3 0x7ff90951bd5c base::WaitableEvent::TimedWaitUntil()
4 0x7ff90951bcb6 base::WaitableEvent::Wait()
5 0x7ff8f8a4cd70 mojo::core::NodeController::WaitSyncSharedBuffer()
6 0x7ff8f8a39134 mojo::core::Core::WaitSyncPlatformSharedMemoryRegion()
7 0x7ff908587418 MojoWaitSyncPlatformSharedMemoryRegion
8 0x7ff90859ec30 mojo::WaitSyncSharedMemory()

Signed-off-by: Suhaspoornachandra <s.poornachan@samsung.com>

suyambulingamrm commented on mojo/core/broker_castanets.cc:
```diff
@@ -246,23 +243,18 @@ void BrokerCastanets::OnBufferSync(uint64_t guid_high, uint64_t guid_low,
-  int fd = base::SharedMemoryTracker::GetInstance()->Find(guid);
-  base::subtle::PlatformSharedMemoryRegion handle;
-  if (fd < 0) {
     base::SharedMemoryCreateOptions options;
```
Looks like indendation is not proper from L 246 to L257 can you cross check once.

suyambulingamrm commented on mojo/core/castanets_fence.h:
```diff
@@ -54,16 +52,16 @@ class CastanetsFence : public base::RefCountedThreadSafe<CastanetsFence> {
 
 class CastanetsFenceManager {
- public:
+public:
```
Are the indentations correct.. ? Can you cross check with 'git cl format'

Commit:
Add .clang format from M85 upstream

Signed-off-by: Suhaspoornachandra <s.poornachan@samsung.com>

Commit:
Remove BUFFER_SYNC_ACK.

The ACK method waits for the sender(writer) to complete the sync.
This has two major problems.
  1) Worst Performance
  2) This cannot be applied to the resource loader,
     because it writes in the IO thread.
Castanets will introduce a fence method which the receiver(reader)
waits for the sync to complete.

Signed-off-by: Suhaspoornachandra <s.poornachan@samsung.com>
